### PR TITLE
Add lm_inference for conditional text generation

### PR DESCRIPTION
**Pull request by pyf98**
This PR adds `lm_inference`, which can be used for conditional text generation.

An example using the LM from `librispeech` is as follows.

Code:
```
import logging
import torch
from espnet2.bin.lm_inference import GenerateText


if __name__ == "__main__":
    logging.basicConfig(
        level=logging.DEBUG,
        format="%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
    )

    generatetext = GenerateText(
        lm_file="exp/asapp/e_branchformer_librispeech/lm/valid.loss.ave_10best.pth",
        token_type="bpe",
        bpemodel="path/to/bpemodel",
        device="cuda" if torch.cuda.is_available() else "cpu",
        maxlen=15,
        minlen=1,
        beam_size=5,
        penalty=0,
    )

    res = generatetext(
        "i have traveled to many ".upper()
    )
    print(res[0][0])
    print(res[0][1])
```

Output:
```
I HAVE TRAVELED TO MANY PARTS OF THE WORLD
['▁I', '▁HAVE', '▁TRAVEL', 'ED', '▁TO', '▁MANY', '▁PARTS', '▁OF', '▁THE', '▁WORLD']
```

The script is ready to use but not finished. I will clean the code soon.

### CODE REVIEW

**Comment by sw005320** [comment]
Sounds very good.
I think it is generally good, but we may start to think of preparing a sampling-based text generation instead of beam search in the future.

@soumimaiti, can you also review this PR?

**Comment by pyf98** [comment]
Yes, we will need sampling instead of searching for generation purposes.
For ASR or TTS, beam search should be sufficient at this moment.

**Comment by pyf98** [comment]
I added an `inference` method so that this script can be executed. A `text` file can be provided as the condition. An example is attached below.

Script:
```bash
python -m espnet2.bin.lm_inference \
    --log_level DEBUG \
    --output_dir exp/tmp_decode \
    --ngpu 1 \
    --data_path_and_name_and_type data/tmp/text,text,text \
    --lm_file exp/asapp/e_branchformer_librispeech/lm/valid.loss.ave_10best.pth \
    --beam_size 5 \
    --penalty 2 \
    --maxlen 30 \
    --minlen 1 \
    --token_type bpe \
    --bpemodel data/en_token_list/bpe_unigram5000/bpe.model
```

Input text file:
```
test_1 I HAVE TRAVELLED TO MANY
test_2 DO YOU KNOW
```

Output text file:
```
test_1 I HAVE TRAVELLED TO MANY PARTS OF THE WORLD BUT I HAVE NEVER MET WITH ANY ONE WHO SEEMED TO ME SO MUCH LIKE A HUMAN BEING AS YOU HAVE BEEN TO ME
test_2 DO YOU KNOW WHAT I'M GOING TO DO WITH YOU
```

**Review by soumimaiti — COMMENTED** [review]
Can we also support temparature-based sampling for inference?

File: espnet2/bin/lm_inference.py [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+        help="Minimum output length",
+    )
+    group.add_argument("--ngram_weight", type=float, default=0.9, help="ngram weight")
+
+    group = parser.add_argument_group("Text converter related")
+    group.add_argument(
+        "--token_type",
+        type=str_or_none,
+        default=None,
+        choices=["char", "bpe", None],
```
[the remark this answers is not in the document]
Comment (pyf98) [reply espnet2/bin/lm_inference.py]: Just added `word` in `token_type`.

**Comment by sw005320** [comment]
LGTM!
Thanks!

**Comment by pyf98** [comment]
Hi @sw005320 Thanks for sharing the info about end detection. I just checked the corresponding code here: https://github.com/espnet/espnet/blob/master/espnet/nets/e2e_asr_common.py#L17

This is used for early stopping during beam search. If recently generated hypotheses have much lower scores than the best hypothesis, then the beam search will terminate before `maxlen`.

I feel this should work in general as long as we are searching for the hypothesis with the highest score. The main issue in generation tasks is that we want to obtain diverse samples instead of the best one. In this case, we probably should not use end detection. Instead, we keep generating tokens until `maxlen` and return them.

**Pull request merged**